core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/AbstractParsedPercentiles.java

```diff
@@ -0,0 +1,169 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.metrics.percentiles;
+
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.ParsedAggregation;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractParsedPercentiles extends ParsedAggregation implements Iterable<Percentile>  {
+
+    private final List<InternalPercentile> percentiles = new ArrayList<>();
+    private final Map<Double, String> percentilesAsString = new HashMap<>();
+
+    private boolean keyed;
+
+    void addPercentile(double key, double value) {
+        percentiles.add(new InternalPercentile(key, value));
+    }
+
+    void addPercentileAsString(double key, String valueAsString) {
+        percentilesAsString.put(key, valueAsString);
+    }
+
+    InternalPercentile getPercentile(double percent) {
+        for (InternalPercentile percentile : percentiles) {
+            if (percentile.getPercent() == percent) {
+                return percentile;
+            }
+        }
+        return null;
+    }
+
+    String percentileAsString(double percent) {
+        return percentilesAsString.get(percent);
+    }
+
+    void setKeyed(boolean keyed) {
+        this.keyed = keyed;
+    }
+
+    @Override
+    public Iterator<Percentile> iterator() {
+        return new Iterator<Percentile>() {
+            final Iterator<InternalPercentile> iterator = percentiles.iterator();
+            @Override
+            public boolean hasNext() {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Percentile next() {
+                return iterator.next();
+            }
+        };
+    }
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject(CommonFields.VALUES.getPreferredName());
+            for (InternalPercentile percentile : percentiles) {
+                Double key = percentile.getPercent();
+                builder.field(String.valueOf(key), percentile.getValue());
+
+                String valueAsString = percentileAsString(key);
+                if (valueAsString != null && valueAsString.isEmpty() == false) {
```

Comment:
Strings.hasLength? But when can it happen that the key is empty? isn't that an issue?